cmd/activator/main.go

```diff
@@ -258,31 +289,28 @@ func main() {
 		logger.Fatalw("Failed to start configuration manager", zap.Error(err))
 	}
 
-	servers := map[string]*http.Server{
-		"http1":   network.NewServer(":"+strconv.Itoa(networking.BackendHTTPPort), ah),
-		"h2c":     network.NewServer(":"+strconv.Itoa(networking.BackendHTTP2Port), ah),
-		"profile": profiling.NewServer(profilingHandler),
-	}
+	servers["http1"] = network.NewServer(":"+strconv.Itoa(networking.BackendHTTPPort), ah)
+	servers["h2c"] = network.NewServer(":"+strconv.Itoa(networking.BackendHTTP2Port), ah)
+	servers["profile"] = profiling.NewServer(profilingHandler)
 
-	errCh := make(chan error, len(servers))
+	eg := errgroup.Group{}
 	for name, server := range servers {
-		go func(name string, s *http.Server) {
+		name, server := name, server
+		eg.Go(func() error {
 			// Don't forward ErrServerClosed as that indicates we're already shutting down.
-			if err := s.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-				errCh <- perrors.Wrapf(err, "%s server failed", name)
+			if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				return perrors.Wrapf(err, "%s server failed", name)
 			}
-		}(name, server)
-	}
-	// Exit as soon as we see a shutdown signal or one of the servers failed.
-	select {
-	case <-ctx.Done():
-	case err := <-errCh:
-		logger.Errorw("Failed to run HTTP server", zap.Error(err))
+			return nil
+		})
 	}
 
-	for _, server := range servers {
-		server.Shutdown(context.Background())
+	if err := eg.Wait(); err != nil {
```

Comment:
If one of the servers crashes, I don't think the Wait will trigger as we'll be waiting for all of them to crash. You can however create an errrgoup from context. That newly created context will trigger once the first function returns an error. Yes, it's nasty.

Reply:
> The first call to return a non-nil error cancels the group; its error will be returned by Wait.

From godoc on `Go()`

Reply:
See

https://github.com/golang/sync/blob/112230192c580c3556b8cee6403af37a4fc5f28c/errgroup/errgroup.go#L39-L45

And the `Wait` doc:

> Wait blocks until all function calls from the Go method have returned, then returns the first non-nil error (if any) from them. 

It's subtle. I think "cancels the group" refers to the context.

Reply:
tl;dr I'm changing this back to what it did before (no errgroup)

The way I'd initially rewritten this, we'd actually never complete the Wait, so the pods stick around for the entire grace period.  I'm reworking this now.